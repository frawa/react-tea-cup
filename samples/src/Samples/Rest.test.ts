/*
 * MIT License
 *
 * Copyright (c) 2019 RÃ©mi Van Keisbelck
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 */

import { view, Msg, init, Model } from './Rest';
import { render, fireEvent } from '@testing-library/react';
import { extendJest, Testing } from 'react-tea-cup';
import { Cmd } from 'tea-cup-core';
import * as matchers from '@testing-library/jest-dom/matchers'

extendJest(expect);
const testing = new Testing<Msg>();
expect.extend(matchers)

describe('Test Rest', () => {
  describe('init state', () => {
    test('triggers message', () => {
      const [state, cmd] = init();
      expect(cmd).not.toEqual(Cmd.none());
    });
  });

  describe('view state', () => {
    const [initialState] = init();

    test('snapshot initial', () => {
      const { container } = render(view(testing.noop, initialState));
      expect(container).toMatchSnapshot();
    });

    test('snapshot loaded', () => {
      const loaded: Model = {
        tag: 'loaded',
        commits: [{ sha: '13131313', author: 'Toto' }],
      };
      const { container } = render(view(testing.noop, loaded));
      expect(container).toMatchSnapshot();
    });

    test('snapshot error', () => {
      const errored: Model = {
        tag: 'load-error',
        error: new Error('boom'),
      };
      const { container } = render(view(testing.noop, errored));
      expect(container).toMatchSnapshot();
    });
  });

  describe('clicking generates messages which update', () => {
    test('click list commits', () => {
      const loaded: Model = {
        tag: 'loaded',
        commits: [{ sha: '13131313', author: 'Toto' }],
      };
      const { container } = render(view(testing.dispatcher, loaded));
      fireEvent.click(container.querySelectorAll('div > button').item(0));

      const [newState, cmd] = testing.dispatched!(loaded);
      expect(newState).toEqual({ tag: 'loading' });
      expect(cmd).not.toEqual(Cmd.none());
    });

    test('click list commits with loading', () => {
      const loaded: Model = {
        tag: 'loaded',
        commits: [{ sha: '13131313', author: 'Toto' }],
      };
      const { container } = render(view(testing.dispatcher, loaded));
      fireEvent.click(container.querySelectorAll('div > button').item(0));

      const [newState, cmd] = testing.dispatched!(loaded);
      expect(newState).toEqual({ tag: 'loading' });
      expect(cmd).not.toEqual(Cmd.none());

      const mockedCommits = [{ sha: '13131313', commit: { author: { name: 'Toto' } } }];

      jest.autoMockOff();
      fetchMock.resetMocks();
      fetchMock.once(JSON.stringify(mockedCommits));

      testing.dispatchFrom(cmd).then((msg: Msg) => {
        const [newState, cmd1] = msg(loaded);
        expect(cmd1).toEqual(Cmd.none());
        expect(newState).toEqual({ tag: 'loaded', commits: [{ sha: '13131313', author: 'Toto' }] });
      });
    });

    test('click errored', () => {
      const errored: Model = {
        tag: 'load-error',
        error: new Error('boom'),
      };
      const { container } = render(view(testing.dispatcher, errored));
      fireEvent.click(container.querySelectorAll('div > button').item(0));

      const [newState, cmd] = testing.dispatched!(errored);
      expect(newState).toEqual({ tag: 'loading' });
      expect(cmd).not.toEqual(Cmd.none());
    });
  });
});
