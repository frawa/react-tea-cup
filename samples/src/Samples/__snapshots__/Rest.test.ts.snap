// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Rest view state snapshot error 1`] = `
<div>
  <div>
    <p>
      This shows how to use Http and decode. Click to list commits using the github API.
    </p>
    <button>
      List commits
    </button>
    <div
      style="min-height: 300px;"
    >
      <p>
        Ooops ! There was an error : 
        boom
      </p>
    </div>
  </div>
</div>
`;

exports[`Test Rest view state snapshot initial 1`] = `
<div>
  <div>
    <p>
      This shows how to use Http and decode. Click to list commits using the github API.
    </p>
    <button
      disabled=""
    >
      Loading...
    </button>
    <div
      style="min-height: 300px;"
    />
  </div>
</div>
`;

exports[`Test Rest view state snapshot loaded 1`] = `
<div>
  <div>
    <p>
      This shows how to use Http and decode. Click to list commits using the github API.
    </p>
    <button>
      List commits
    </button>
    <div
      style="min-height: 300px;"
    >
      <ul>
        <li>
          13131313
           - 
          Toto
        </li>
      </ul>
    </div>
  </div>
</div>
`;
